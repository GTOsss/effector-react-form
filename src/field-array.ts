import React, {useCallback, useEffect, useMemo} from 'react';
import {useStore} from 'effector-react';
import {
  FieldArrayParams, MapFieldArrayCallback,
} from '../index';
import {getIn, setIn, removeFromInlineMap} from './utils/object-manager';
import {createEvent} from 'effector';

const useFieldArray = <Values>({$fieldsInline, $values, name}: FieldArrayParams<Values>) => {
  const remove = useMemo(() => createEvent<number>('hookForm_fieldArray_Remove'), []);
  const push = useMemo(() => createEvent<any>('hookForm_fieldArray_Push'), []);
  const unshift = useMemo(() => createEvent<any>('hookForm_fieldArray_Unshift'), []);
  const move = useMemo(() => createEvent<{from: number, to: number}>('hookForm_fieldArray_Move'), []);
  const insert = useMemo(() => createEvent<{value: any, index: number}>('hookForm_fieldArray_Insert'), []);
  const swap = useMemo(() => createEvent<{from: number, to: number}>('hookForm_fieldArray_Swap'), []);

  const values = useStore($values);

  const map = useCallback((callback: MapFieldArrayCallback) => {
    const results = [];
    const fields = getIn(values, name, []);
    fields.forEach((field, index) => {
      const callbackResult = callback({
        formItemName: `${name}.${index}`,
        fields,
        field,
        index,
      });
      results.push(callbackResult);
    });
    return results;
  }, [values]);

  useEffect(() => {
    $fieldsInline.on(remove, (fieldsInline, index) => removeFromInlineMap(fieldsInline, `${name}.${index}`));
    $values.on(remove, (values, index) => {
      const newFields = getIn(values, name, []).filter((_, i) => i !== index);
      return setIn(values, name, newFields);
    });

    // $fieldsInline will be initialize automatically because it's new field
    $values.on(push, (values, value) => {
      const newFields = [...getIn(values, name, []), value];
      return setIn(values, name, newFields);
    });

    $values.on(unshift, (values, value) => { // todo implement for fieldsInline
      const newFields = [value, ...getIn(values, name, [])];
      return setIn(values, name, newFields);
    });

    $values.on(move, (values, {from, to}) => { // todo implement for fieldsInline
      const fields = getIn(values, name, []);
      const newFields = [];
      let movingField = {};
      fields.forEach((field, i) => {
        if (from === i) {
          movingField = field;
        } else if (to === i) {
          newFields.push(field);
          newFields.push(movingField);
        } else {
          newFields.push(field);
        }
      });
      return setIn(values, name, newFields);
    });

    $values.on(insert, (values, {index, value}) => { // todo implement for fieldsInline
      const fields = getIn(values, name);
      const newFields = [];
      fields.forEach((field, i) => {
        if (index === i) {
          newFields.push(value);
          newFields.push(field);
        } else {
          newFields.push(field);
        }
      });
      return setIn(values, name, newFields);
    });

    $values.on(swap, (values, {from, to}) => { // todo implement for fieldsInline
      const fields = getIn(values, name);
      const newFields = [];
      fields.forEach((field, i) => {
        if (from === i) {
          newFields.push(fields[to]);
        } else if (to === i) {
          newFields.push(fields[from]);
        } else {
          newFields.push(field);
        }
      });
      return setIn(values, name, newFields);
    });

    return () => {
      $values.off(remove);
      $values.off(push);
      $values.off(unshift);
      $values.off(move);
      $values.off(insert);
      $values.off(swap);
    };
  }, []);

  return {
    map,
    remove,
    push,
    unshift,
    move,
    insert,
    swap,
  };
};

export default useFieldArray;
