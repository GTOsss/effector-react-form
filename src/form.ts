import {useCallback, useEffect, useMemo, SyntheticEvent, useRef} from 'react';
import {
  ControllerHof,
  OnSubmit,
  ControllerInjectedResult,
  FieldState,
  FormState,
  Messages,
  ResultHook,
  UseFormParams, AnyState, ErrorsInline, FieldsInline, SetOrDeleteErrorParams,
} from '../index';
import {createStore, createEvent} from 'effector';
import {useStore} from 'effector-react';
import {getValue} from './utils/dom-helper';
import {setIn, getIn, deleteIn, mapInlineToMapNested} from './utils/object-manager';

const initialFieldState: FieldState = {
  touched: false,
  changed: false,
  blurred: false,
  active: false,
};

const initialFormState = {
  submitted: false,
  hasError: false,
  forcedError: false,
};

const useForm = <Values extends AnyState>({
  $values: $valuesProp,
  $errorsInline: $errorsInlineProp,
  $fieldsInline: $fieldsInlineProp,
  $form: $formProp,
  // validate,
}: UseFormParams<Values> = {}): ResultHook<Values> => {

  const willMount = useRef(true);

  const validateMapByNameRef = useRef<Record<string, any>>({});
  validateMapByNameRef.current = {};
  const setValue = useMemo(() => createEvent<{field: string, value: any}>(`hookForm_SetValue`), []);
  const setOrDeleteError = useMemo(() => createEvent<SetOrDeleteErrorParams>(`hookForm_SetError`), []);
  const setErrorsInlineState = useMemo(() => createEvent<ErrorsInline>(`hookForm_SetErrorsInlineState`), []);
  const setFieldState = useMemo(() => createEvent<{field: string, state: FieldState}>(`hookForm_SetFieldState`), []);
  const setSubmitted = useMemo(() => createEvent<boolean>(`hookForm_SetSubmitted`), []);

  const $values = useMemo(() => $valuesProp || createStore<Values>({} as Values), []);
  const $errorsInline = useMemo(() => $errorsInlineProp || createStore<ErrorsInline>({}), []);
  const $errors = useMemo(() => $errorsInline.map((state) => mapInlineToMapNested<Messages<Values>>(state)), []);
  const $fieldsInline = useMemo(() => $fieldsInlineProp || createStore<FieldsInline>({}), []);
  const $form = useMemo(() => $formProp || createStore<FormState>(initialFormState), []);

  if (willMount.current) {
    $fieldsInline.on(setFieldState, (state, {field, state: fieldState}) => {
      return ({...state, [field]: fieldState});
    });
  }

  const validateForm = useCallback(() => {
    const values = $values.getState();
    const errorsInlineState = {};
    Object.entries(validateMapByNameRef.current).forEach(([name, validate]) => {
      const error = validate && validate(getIn(values, name));
      if (error) {
        errorsInlineState[name] = validate && validate(getIn(values, name));
      }
    });
    setErrorsInlineState(errorsInlineState);
  }, []);

  const values = useStore<Values>($values);

  useEffect(() => {
    validateForm();
  }, [values]);

  useEffect(() => {
    $values.on(setValue, (state, {field, value}) => setIn(state, field, value));

    $errorsInline.on(setOrDeleteError, (state, {field, error}) =>
      error ? {...state, [field]: error} : deleteIn(state, field, false, false));

    $errorsInline.on(setErrorsInlineState, (_, errorsInline) => errorsInline);

    $form.on(setOrDeleteError, (state, {forced = true}) => ({...state, forcedError: forced}));

    $form.on(setSubmitted, (state, value) => setIn(state, 'submitted', value));

    $form.on($errorsInline.updates, (state, errorsInline) =>
      setIn(state, 'hasError', Boolean(Object.keys(errorsInline).length)));

    validateForm();

    return () => {
      $values.off(setValue);

      $errorsInline.off(setOrDeleteError);
      $errorsInline.off(setErrorsInlineState);

      $form.off(setOrDeleteError);
      $form.off(setSubmitted);
      $form.off($errorsInline);
    };
  }, []);

  const controller = useCallback<ControllerHof>(({
    name: nameProp,
    validate,
  }) => {
    validateMapByNameRef.current[nameProp] = validate;

    return (): ControllerInjectedResult => {
      const refName = useRef<string>(nameProp);
      refName.current = nameProp;

      const onChangeBrowser = useMemo(() => createEvent<any>(`hookForm_OnChange_${refName.current}`), []);
      const onChange = useMemo(() => onChangeBrowser.map((eventOrValue) => getValue(eventOrValue)), []);
      const onFocusBrowser = useMemo(() => createEvent<any>(`hookForm_OnFocus_${refName.current}`), []);
      const onBlurBrowser = useMemo(() => createEvent<any>(`hookForm_OnBlur_${refName.current}`), []);

      useEffect(() => {

        $values.on(onChange, (state, value) => setIn(state, refName.current, value));

        $fieldsInline.on(onFocusBrowser, (state) => {
          return {...state, [refName.current]: {...state[refName.current], touched: true, active: true}};
        });
        $fieldsInline.on(onChangeBrowser, (state) => {
          return {...state, [refName.current]: {...state[refName.current], changed: true}};
        });
        $fieldsInline.on(onBlurBrowser, (state) => {
          return {...state, [refName.current]: {...state[refName.current], blurred: true, active: false}};
        });

        setFieldState({field: refName.current, state: initialFieldState});

        return () => {
          $values.off(onChange);

          $fieldsInline.off(onFocusBrowser);
          $fieldsInline.off(onChangeBrowser);
          $fieldsInline.off(onBlurBrowser);
        };
      }, []);

      const values = useStore<Values>($values);
      const errorsInline = useStore<ErrorsInline>($errorsInline);

      const value = getIn(values, nameProp);
      const error = errorsInline[nameProp];

      const fieldsState = useStore<FieldsInline>($fieldsInline);
      const fieldState = fieldsState[nameProp] || initialFieldState;

      const formState = useStore<FormState>($form);

      return {
        input: {
          name: nameProp,
          value,
          onChange: onChangeBrowser,
          onFocus: onFocusBrowser,
          onBlur: onBlurBrowser,
        },
        form: formState,
        fieldState,
        error,
        validate,
        setFieldState,
        setOrDeleteError,
      };
    };
  }, []);

  const handleSubmit = (onSubmit: OnSubmit<Values>) => (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    setSubmitted(true);

    validateForm();

    onSubmit({
      values: $values.getState(),
      errors: $errors.getState(),
      errorsInline: $errorsInline.getState(),
      fieldsInline: $fieldsInline.getState(),
      form: $form.getState(),
    });
  };

  willMount.current = false;

  return {
    controller,
    handleSubmit,
    setValue,
    setOrDeleteError,
    setErrorsInlineState,
    $values,
    $errorsInline,
    $errors,
    $fieldsInline,
    $form,
  };
};

export default useForm;
