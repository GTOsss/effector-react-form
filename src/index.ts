import useForm from './use-form';
import useFieldArray from './use-field-array';
import useError from './use-error';
import createFieldArray from './factories/create-field-array';
import createForm from './factories/create-form';
import { deleteIn, getIn, setIn, makeNested, removeFromInlineMap } from './utils/object-manager';

export * from './ts';
export {
  createForm,
  createFieldArray,
  useForm,
  useError,
  useFieldArray,
  deleteIn,
  getIn,
  setIn,
  makeNested,
  removeFromInlineMap,
};
