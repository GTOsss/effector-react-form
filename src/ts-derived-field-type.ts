type IsRecordLike<T> = T extends object ? (string extends keyof T ? true : false) : false;

export type DerivedFieldType<T> = IsRecordLike<T> extends true
  ? string
  : T extends object
  ? {
      [K in keyof T]: K extends string | number
        ? T[K] extends object
          ? K | `${K}.${DerivedFieldType<T[K]>}`
          : K
        : never;
    }[keyof T]
  : never;
