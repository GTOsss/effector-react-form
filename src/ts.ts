import { Store, Event, Domain } from 'effector';
import React from 'react';

export type AnyState = Record<string, any>;

export type ErrorsInline = Record<string, Message>;

export type FieldsInline = Record<string, FieldState>;

export type Message = string | null | undefined;

export type Messages<Values> = {
  [key in keyof Values]?: Values[key] extends AnyState ? Messages<Values[key]> : Message;
};

export type SubmitParams<Values = any, Meta = any> = {
  values: Values;
  errorsInline: ErrorsInline;
  fieldsInline: FieldsInline;
  form: FormState;
  meta: Meta;
};

export type FormState = {
  submitted: boolean;
  hasError: boolean;
  hasOuterError: boolean;
};

export type FieldState = {
  _type: 'fieldMeta';
  active: boolean;
  touched: boolean;
  changed: boolean;
  blurred: boolean;
  touchedAfterOuterError: boolean;
  changedAfterOuterError: boolean;
  blurredAfterOuterError: boolean;
  validate?: (value: any) => string | undefined;
};

export type ControllerParams = {
  name: string;
  validate?: (value: any) => Message;
};

export type SetValueParams = {
  field: string;
  value: any;
};

export type SetOrDeleteErrorParams = {
  field: string;
  error?: Message;
};

export type SetFieldStateParams = {
  field: string;
  state: FieldState;
};

export type SetOrDeleteOuterErrorParams = {
  field: string;
  error: Message;
};

export type FieldInitParams = {
  name: string;
  validate?: ControllerParams['validate'];
};

export type ControllerInjectedResult<Meta = any> = {
  input: {
    name: string;
    value;
    onChange: (event: any) => void;
    onFocus: (event: any) => void;
    onBlur: (event: any) => void;
  };
  error: Message;
  innerError: Message;
  outerError: Message;
  isShowError: boolean;
  isShowOuterError: boolean;
  isShowInnerError: boolean;
  form: FormState;
  validate?: (value: any) => Message;
  setFieldState: ({ field: string, state: FieldState }) => void;
  setOrDeleteError: ({ field: string, error: Message }) => void;
  setOrDeleteOuterError: ({ field: string, error: Message }) => void;
  setOuterErrorsInlineState: (errors: ErrorsInline) => void;
  fieldState: FieldState;
};

export type Controller<Meta = any> = () => ControllerInjectedResult<Meta>;

export type ControllerHof<Meta = any> = (a: ControllerParams) => Controller<Meta>;

export type FormValidateParams<Values> = {
  values: Values;
  errorsInline: ErrorsInline;
};

export type FormValidate<Values> = (params: FormValidateParams<Values>) => ErrorsInline;

export type MapSubmit<Values, ResultValues, Meta = any> = (
  params: SubmitParams<Values, Meta>,
) => SubmitParams<ResultValues, Meta>;

export type OnSubmit<Values, Meta = any> = (params: SubmitParams<Values, Meta>) => void;

export type OnChange<Values, Meta = any> = OnSubmit<Values, Meta>;

export type UseErrorParams<Values> = {
  name: string;
  form: Form<Values>;
};

export type UseErrorResult<Meta = any> = {
  inputValue: any;
  form: FormState;
  fieldState: FieldState;
  error: Message;
  innerError: Message;
  outerError: Message;
  isShowError: boolean;
  isShowOuterError: boolean;
  isShowInnerError: boolean;
};

export type FieldArrayParams<Values> = {
  name: string;
  fieldArray: FieldArray<Values>;
};

export type MapFieldsArrayCallbackParams = {
  formItemName: string;
  field: any;
  fields: Array<any>;
  index: number;
};

export type MapFieldArrayCallback = (params: MapFieldsArrayCallbackParams) => React.ReactNode;

export type ResultUseFieldArray = {
  map: (fn: MapFieldArrayCallback) => React.ReactNode[];
  remove: (index: number) => void;
  push: (value: any | Array<any>) => void;
};

export type CreateFormParams<Values = any, MappedValues = any, Meta = any> = {
  validate?: FormValidate<Values>;
  mapSubmit?: MapSubmit<Values, MappedValues, Meta>;
  onSubmit?: OnSubmit<Values, Meta>;
  onChange?: OnChange<Values, Meta>;
  initialValues?: Values;
  initialMeta?: Meta;
  domain?: Domain;
};

export type Form<Values = any, Meta = any> = {
  $values: Store<Values>;
  $errorsInline: Store<ErrorsInline>;
  $outerErrorsInline: Store<ErrorsInline>;
  $form: Store<FormState>;
  $fieldsInline: Store<Record<string, FieldState>>;

  setValue: Event<any>;
  setOrDeleteError: Event<any>;
  setErrorsInlineState: Event<any>;
  setFieldState: Event<any>;
  setSubmitted: Event<any>;
  resetOuterFieldStateFlags: Event<any>;
  setOrDeleteOuterError: Event<any>;
  reset: Event<any>;

  setMeta: Event<any>;

  setOuterErrorsInlineState: Event<any>;
  validateForm: Event<any>;
  submit: Event<any>;
  onSubmit: Event<SubmitParams<Values, Meta>>;

  onChangeFieldBrowser: Event<{ event: React.SyntheticEvent; name: string }>;
  onFocusFieldBrowser: Event<{ event: React.SyntheticEvent; name: string }>;
  onBlurFieldBrowser: Event<{ event: React.SyntheticEvent; name: string }>;
  fieldInit: Event<{ name: string; validate?: ControllerParams['validate'] }>;
};

export type FieldArray<Values> = {
  form: Form<Values>;
  push: Event<{ fieldName: string; value: any | any[] }>;
  remove: Event<{ fieldName: string; index: number }>;
};

export type CreateFieldArrayParams<Values> = {
  form: Form<Values>;
  domain?: Domain;
};

// declare const useFieldArray: <Values extends AnyState = AnyState>(
//   paramsFieldArray: FieldArrayParams,
// ) => ResultUseFieldArray<Values>;
//
// declare const setIn: <O extends AnyState = AnyState, V = any>(
//   object: O,
//   path: string,
//   value: V,
// ) => O;
//
// declare const deleteIn: <O extends AnyState = AnyState>(
//   object: O,
//   path: string,
//   removeEmpty: boolean = false,
//   inDeep: boolean = true,
// ) => O;
//
// declare const getIn: <O extends AnyState = AnyState, V = any>(
//   object: O,
//   path: string,
//   removeEmpty: boolean = false,
//   inDeep: boolean = true,
// ) => V = any;
//
// declare const makeNested: <O extends AnyState = AnyState, R extends AnyState>(
//   inlineMap: O,
// ) => R;
//
// declare const removeFromInlineMap: <O extends FieldsInline = FieldsInline, R extends FieldsInline = FieldsInline>(
//   inlineMap: O,
//   key: string,
// ) => R;
