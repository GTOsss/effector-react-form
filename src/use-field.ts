import React, { useCallback, useEffect } from 'react';
import { useEvent, useStore } from 'effector-react';
import {
  FieldState,
  FormState,
  AnyState,
  Message,
  ControllerFieldHof,
  ControllerFieldInjectedResult,
  Field,
} from './ts';

type UseFieldParamsWithFactory<Value extends object, Meta> = {
  field: Field<Value>;
  meta?: Meta;
  resetUnmount?: boolean;
};

type UseFieldResultWithFactory<Value> = {
  controller: ControllerFieldHof;
  handleSubmit: (e: React.SyntheticEvent<HTMLFormElement>) => void;
  setMeta: (meta: any) => void;
  setValue: (params: Value) => void;
  setOrDeleteError: (error: Message) => void;
  setFieldState: (fieldState: FieldState) => void;
  setOrDeleteOuterError: (error: Message) => void;
  validateForm: () => void;
  submit: (e: any) => void;
  reset: () => void;
};

const useField = <Value extends AnyState = AnyState, Meta = any>({
  field,
  meta,
  resetUnmount = true,
}: UseFieldParamsWithFactory<Value, Meta>): UseFieldResultWithFactory<Value> => {
  const { $value, $field, $fieldInline, $error, $outerError } = field;

  const setMeta = useEvent(field.setMeta);
  const setValue = useEvent(field.setValue);

  const setOrDeleteError = useEvent(field.setOrDeleteError);
  const setFieldState = useEvent(field.setFieldState);
  const setOrDeleteOuterError = useEvent(field.setOrDeleteOuterError);

  const validateForm = useEvent(field.validateField) as any;
  const submit = useEvent(field.submit) as any;
  const reset = useEvent(field.reset) as any;

  const onChangeFieldBrowser = useEvent(field.onChangeFieldBrowser);
  const onFocusFieldBrowser = useEvent(field.onFocusFieldBrowser);
  const onBlurFieldBrowser = useEvent(field.onBlurFieldBrowser);

  useEffect(() => {
    validateForm();

    return () => {
      if (resetUnmount) {
        reset();
      }
    };
  }, []);

  useEffect(() => {
    setMeta(meta);
  }, [meta]);

  const controller = useCallback<ControllerFieldHof>(() => {
    return (): ControllerFieldInjectedResult => {
      const value = useStore($value);

      const innerError = useStore($error);

      const outerError = useStore($outerError);
      const error = outerError || innerError;

      const fieldInline = useStore<FieldState>($fieldInline);

      const { submitted: fieldSubmitted } = useStore<FormState>($field);

      const { hasError: fieldHasError } = useStore<FormState>($field);
      const { hasOuterError: fieldHasOuterError } = useStore<FormState>($field);
      const fieldState: FormState = {
        submitted: fieldSubmitted,
        hasError: fieldHasError,
        hasOuterError: fieldHasOuterError,
      };

      const isShowInnerError = (fieldState.submitted || fieldInline.blurred) && Boolean(innerError);
      const isShowOuterError = !fieldInline.changedAfterOuterError && Boolean(outerError);
      const isShowError = isShowInnerError || isShowOuterError;

      return {
        input: {
          value,
          onChange: (event) => onChangeFieldBrowser({ event }),
          onFocus: (event) => onFocusFieldBrowser({ event }),
          onBlur: (event) => onBlurFieldBrowser({ event }),
        },
        field: fieldState,
        meta,
        fieldInline,
        error,
        innerError,
        outerError,
        isShowError,
        isShowOuterError,
        isShowInnerError,
        setFieldState,
        setOrDeleteError,
        setOrDeleteOuterError,
      };
    };
  }, []);

  const handleSubmit = useCallback((e: React.SyntheticEvent<HTMLFormElement>): void => {
    e.preventDefault();
    submit();
  }, []);

  return {
    controller,
    handleSubmit,
    setMeta,
    setValue,
    setOrDeleteError,
    setFieldState,
    setOrDeleteOuterError,
    validateForm,
    submit,
    reset,
  };
};

export default useField;
