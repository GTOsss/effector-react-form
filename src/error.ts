import {useStoreMap} from 'effector-react';
import {ErrorsInline, FieldsInline, FieldState, FormState, Message, UseErrorParams} from '../index';
import {getIn} from './utils/object-manager';
import {initialFieldState} from './form';

const useError = <Values>({
  name,
  $values,
  $errorsInline,
  $outerErrorsInline,
  $form,
  $fieldsInline,
}: UseErrorParams) => {
  const value = useStoreMap<Values, any, [string]>({
    store: $values,
    keys: [name],
    fn: (values, [field]) => getIn(values, field) || null,
  });

  const innerError = useStoreMap<ErrorsInline, Message, [string]>({
    store: $errorsInline,
    keys: [name],
    fn: (errorsInline, [field]) => errorsInline[field] || null,
  });
  const outerError = useStoreMap<ErrorsInline, Message, [string]>({
    store: $outerErrorsInline,
    keys: [name],
    fn: (outerErrorsInline, [field]) => outerErrorsInline[field] || null,
  });
  const error = innerError || outerError;

  const fieldState = useStoreMap<FieldsInline, FieldState, [string]>({
    store: $fieldsInline,
    keys: [name],
    fn: (fieldsInline, [field]) => fieldsInline[field] || initialFieldState,
  });

  const formSubmitted = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.submitted,
  });
  const formHasError = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.hasError,
  });
  const formHasOuterError = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.hasOuterError,
  });
  const formState: FormState = {
    submitted: formSubmitted,
    hasError: formHasError,
    hasOuterError: formHasOuterError,
  };

  const isShowInnerError = (formState.submitted || fieldState.blurred) && Boolean(innerError);
  const isShowOuterError = (!fieldState.changedAfterOuterError) && Boolean(outerError);
  const isShowError = isShowInnerError || isShowOuterError;

  return {
    inputValue: value,
    form: formState,
    fieldState,
    error,
    innerError,
    outerError,
    isShowError,
    isShowOuterError,
    isShowInnerError,
  };
};

export default useError;
