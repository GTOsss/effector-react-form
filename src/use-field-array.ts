import { ReactNode, useCallback, useRef, useMemo } from 'react';
import { useUnit } from 'effector-react';
import { AnyState, FieldArrayParams, MapFieldArrayCallback, ResultUseFieldArray } from './ts';
import { getIn } from './utils/object-manager';

const useFieldArray = <Values extends object = AnyState>({
  fieldArray,
  name,
}: FieldArrayParams<Values>): ResultUseFieldArray => {
  const refName = useRef(name);
  refName.current = name;

  const {
    form: { $values },
  } = fieldArray;

  const push = useUnit(fieldArray.push);
  const remove = useUnit(fieldArray.remove);

  const values = useUnit($values);

  const map = useCallback<(fn: MapFieldArrayCallback) => ReactNode[]>(
    (callback: MapFieldArrayCallback) => {
      const results = [];
      const fields = getIn(values, refName.current, []) as any[];
      fields.forEach((field, index) => {
        const callbackResult = callback({
          formItemName: `${refName.current}.${index}`,
          fields,
          field,
          index,
        });
        results.push(callbackResult);
      });
      return results;
    },
    [values],
  ); // todo Fix type

  const count = useMemo(() => (getIn(values, refName.current, []) as any[]).length, [values]);

  return {
    map,
    remove: (index: number) => remove({ fieldName: refName.current, index }),
    push: (value: any | any[]) => push({ fieldName: refName.current, value }),
    count,
  };
};

export default useFieldArray;
