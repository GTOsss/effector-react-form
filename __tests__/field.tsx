import { render, screen, fireEvent } from '@testing-library/react';
import { createStore, sample } from 'effector';
import React from 'react'
import { ControllerField, createField, useField } from "../src";
import { initialFieldState } from '../src/default-states';

interface InputProps {
  controller: ControllerField;
  label: string;
}

const validateLength = ({ value }) => {
  if (value && value.length < 3) {
    return 'Min length is 3 characters'
  } else if (!value) {
    return 'Field is required'
  }
  return undefined;
}

const Input: React.FC<InputProps> = ({ controller, label }) => {
  const { input } = controller();

  return (
    <div role="wrapper-for-input" className="input-wrap">
      <label>{label}</label>
      <input
        {...input}
        value={input.value || ''}
        role="textbox"
        className="input"
        placeholder={label}
      />
    </div>
  );
};

const Field = ({ field }) => {
  const { handleSubmit, controller, submit } = useField({ field });

  return (
    <form onSubmit={handleSubmit}>
      <Input label="Username" controller={controller()} />
      <button onClick={submit}>
        submit
      </button>
    </form>
  );
};

describe('Single field', () => {
  test('submit', () => {
    const field = createField();
    render(<Field field={field} />);
    const buttonSubmit = screen.getByText('submit')
    fireEvent.click(buttonSubmit);
    expect(field.$field.getState().submitted).toEqual(true);
  })

  test('submit snapshot', () => {
    const field = createField();
    render(<Field field={field} />);
    const buttonSubmit = screen.getByText('submit')
    fireEvent.click(buttonSubmit);
    const inputs = screen.getAllByRole('wrapper-for-input');
    expect(inputs).toMatchSnapshot();
  })

  test('onChange', () => {
    const field = createField();
    render(<Field field={field} />);
    const input = screen.getByPlaceholderText('Username');
    fireEvent.change(input, { target: { value: 'login' }});
    expect(field.$value.getState()).toBe('login');
  })

  test('validate minLength', () => {
    const field = createField({ validate: validateLength });
    render(<Field field={field} />);
    const input = screen.getByPlaceholderText('Username');
    const buttonSubmit = screen.getByText('submit')

    fireEvent.click(buttonSubmit);
    expect(field.$error.getState()).toEqual('Field is required');

    fireEvent.change(input, { target: { value: 'lo' } });
    fireEvent.click(buttonSubmit);
    expect(field.$error.getState()).toEqual('Min length is 3 characters');
  })

  test('Field inline', () => {
    const field = createField({ validate: validateLength });
    render(<Field field={field} />);

    expect(field.$fieldInline.getState()).toEqual(initialFieldState);

    const input = screen.getByPlaceholderText('Username');
    fireEvent.change(input, { target: { value: 'name' }});

    expect(field.$fieldInline.getState()).toEqual({ ...initialFieldState, changed: true });

    fireEvent.focus(input);
    expect(field.$fieldInline.getState()).toEqual({ ...initialFieldState, changed: true, touched: true, active: true });

    fireEvent.focusOut(input);
    expect(field.$fieldInline.getState()).toEqual({ ...initialFieldState, changed: true, touched: true, active: false, blurred: true });
  })

  test('Outer validation', () => {
    const field = createField({ validate: validateLength });
    const $outerErrorText = createStore('Outer error');

    sample({
      source: $outerErrorText,
      clock: field.onSubmit,
      target: field.setOrDeleteOuterError,
    })

    render(<Field field={field} />);

    expect(field.$fieldInline.getState()).toEqual(initialFieldState);

    const input = screen.getByPlaceholderText('Username');
    const buttonSubmit = screen.getByText('submit')

    fireEvent.click(buttonSubmit);
    expect(field.$outerError.getState()).toEqual('Outer error');

    
    fireEvent.focus(input);
    expect(field.$fieldInline.getState()).toEqual({
      ...initialFieldState,
      changed: false,
      touched: true,
      active: true,
      blurred: false,
      touchedAfterOuterError: true,
    });

    fireEvent.change(input, { target: { value: 'name' } });

    expect(field.$fieldInline.getState()).toEqual({
      ...initialFieldState,
      active: true,
      changed: true,
      touched: true,
      changedAfterOuterError: true,
      touchedAfterOuterError: true
    });


    fireEvent.focusOut(input);
    expect(field.$fieldInline.getState()).toEqual({
      ...initialFieldState,
      changed: true,
      touched: true,
      active: false,
      blurred: true,
      changedAfterOuterError: true,
      touchedAfterOuterError: true,
      blurredAfterOuterError: true
    });
  })
})