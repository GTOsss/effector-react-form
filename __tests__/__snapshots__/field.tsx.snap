// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single field submit snapshot 1`] = `
Array [
  <div
    class="input-wrap"
    role="wrapper-for-input"
  >
    <label>
      Username
    </label>
    <input
      class="input"
      placeholder="Username"
      role="textbox"
      value=""
    />
  </div>,
]
`;
